tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java

```diff
@@ -0,0 +1,296 @@
+                    selector,
+                    metadata,
+                    clientId,
+                    DEFAULT_MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION,
+                    DEFAULT_RECONNECT_BACKOFF_MS,
+                    DEFAULT_RECONNECT_BACKOFF_MAX,
+                    DEFAULT_SEND_BUFFER_BYTES,
+                    DEFAULT_RECEIVE_BUFFER_BYTES,
+                    config.getInt(CommonClientConfigs.REQUEST_TIMEOUT_MS_CONFIG),
+                    config.getLong(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG),
+                    config.getLong(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG),
+                    time,
+                    true,
+                    new ApiVersions(),
+                    logContext,
+                    MetadataRecoveryStrategy.NONE);
+            ConsumerNetworkClient highLevelClient = new ConsumerNetworkClient(
+                    logContext,
+                    networkClient,
+                    metadata,
+                    time,
+                    config.getLong(CommonClientConfigs.RETRY_BACKOFF_MS_CONFIG),
+                    config.getInt(CommonClientConfigs.REQUEST_TIMEOUT_MS_CONFIG),
+                    Integer.MAX_VALUE);
+            return new AdminClient(time, highLevelClient, metadata.fetch().nodes());
+        }
+
+        AdminClient(Time time, ConsumerNetworkClient client, List<Node> bootstrapBrokers) {
+            this.time = time;
+            this.client = client;
+            this.bootstrapBrokers = bootstrapBrokers;
+        }
+
+        private AbstractResponse send(Node target, AbstractRequest.Builder<?> request) {
+            RequestFuture<ClientResponse> future = client.send(target, request);
+            while (!future.isDone()) {
+                client.poll(time.timer(DEFAULT_REQUEST_TIMEOUT_MS));
+            }
+            if (future.succeeded()) {
+                return future.value().responseBody();
+            } else {
+                throw future.exception();
+            }
+        }
+
+        private AbstractResponse sendAnyNode(AbstractRequest.Builder<?> request) {
+            for (Node broker : bootstrapBrokers) {
+                try {
+                    return send(broker, request);
+                } catch (AuthenticationException e) {
+                    throw e;
+                } catch (Exception e) {
+                    LOGGER.debug("Request {} failed against node {}", request.apiKey(), broker, e);
+                }
+            }
+            throw new RuntimeException("Request " + request.apiKey() + " failed on brokers " + bootstrapBrokers);
+        }
+
+        private KafkaFuture<NodeApiVersions> getNodeApiVersions(Node node) {
+            final KafkaFutureImpl<NodeApiVersions> future = new KafkaFutureImpl<>();
+            try {
+                ApiVersionsResponse response = (ApiVersionsResponse) send(node, new ApiVersionsRequest.Builder());
```

Comment:
That will send the API_VERSION request to brokers one by one. Could you please leverage the `Future` (or `CompletableFuture`) to speedup it?

Reply:
Hi @chia7712, original class sends request one by one. I think we can keep the behavior and do some refactor in the future if someone complains about this. Thanks.

https://github.com/apache/kafka/blob/c5e91546721b6b160732667ca7450d22927cc3b3/core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala#L136-L138